File: src/deadlock_build_sync/policy_model.py
from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum

from .policy_cards import (  # ruff: ignore[typing-only-first-party-import]
    CoreAlternativeCard,
    CounterCard,
)
from .policy_claims import EvidenceClaim, PolicyError
from .policy_graph import (
    NodeKind,
    PolicyNode,
)
from .snapshot import sha256_json


class AbstentionReason(StrEnum):
    STALE_MECHANICS = "stale_or_incomplete_mechanics"
    INADEQUATE_SUPPORT = "inadequate_support_or_overlap"
    EVIDENCE_CONFLICT = "evidence_conflict"
    ILLEGAL_PATH = "illegal_path"
    UNCLEAR_THREAT = "unclear_threat"
    OUT_OF_DISTRIBUTION = "out_of_distribution_state"
    TELEMETRY_FAILURE = "telemetry_failure"


@dataclass(frozen=True)
class Abstention:
    reason: AbstentionReason
    detail: str
    node_id: str | None = None

    def __post_init__(self) -> None:
        """Require an actionable abstention explanation.

        Raises:
            PolicyError: If no detail is supplied.

        """
        if not self.detail.strip():
            raise PolicyError("abstention detail must not be empty")


def _validate_counter_cards(
    cards: tuple[CounterCard, ...],
    nodes: tuple[PolicyNode, ...],
    claim_ids: list[str],
) -> None:
    item_ids = [card.item_id for card in cards]
    if len(item_ids) != len(set(item_ids)):
        raise PolicyError("counter cards must use distinct items")
    claims = set(claim_ids)
    actions = {
        (node.item_id, node.evidence_ref)
        for node in nodes
        if node.kind == NodeKind.PURCHASE and node.optional
    }
    for card in cards:
        if card.evidence_ref not in claims:
            raise PolicyError(
                f"counter card references missing evidence {card.evidence_ref}"
            )
        if (card.item_id, card.evidence_ref) not in actions:
            raise PolicyError("counter card has no matching optional purchase node")


def _validate_core_alternative_cards(
    cards: tuple[CoreAlternativeCard, ...],
    nodes: tuple[PolicyNode, ...],
    claim_ids: list[str],
) -> None:
    item_ids = [card.item_id for card in cards]
    if len(item_ids) != len(set(item_ids)):
        raise PolicyError("core alternative cards must use distinct items")
    claims = set(claim_ids)
    default_items = {
        node.item_id
        for node in nodes
        if node.kind == NodeKind.PURCHASE and not node.optional
    }
    for card in cards:
        if card.evidence_ref not in claims:
            raise PolicyError("core alternative card references missing evidence")
        if (
            card.item_id in default_items
            or card.comparator_item_id not in default_items
        ):
            raise PolicyError("core alternative card does not replace a default item")


@dataclass(frozen=True)
class BuildPolicy:
    schema_version: int
    hero_id: int
    variant: str
    invariant_kit_id: str
    strategic_role: str
    snapshot_id: str
    entry: str
    nodes: tuple[PolicyNode, ...]
    evidence: tuple[EvidenceClaim, ...]
    ability_plan: tuple[PolicyNode, ...] = ()
    abstentions: tuple[Abstention, ...] = ()
    counter_cards: tuple[CounterCard, ...] = ()
    core_alternatives: tuple[CoreAlternativeCard, ...] = ()
    path_id: str = "default"
    path_label: str = "Evidence Default"

    def __post_init__(self) -> None:
        """Validate policy-level identity and uniqueness.

        Raises:
            PolicyError: If identity fields or node/claim IDs are invalid.

        """
        if self.schema_version not in {1, 2, 3, 4, 5}:
            raise PolicyError(f"unsupported policy schema {self.schema_version}")
        if self.hero_id <= 0:
            raise PolicyError("policy hero id must be positive")
        if not all(
            value.strip()
            for value in (
                self.variant,
                self.invariant_kit_id,
                self.strategic_role,
                self.snapshot_id,
                self.entry,
                self.path_id,
                self.path_label,
            )
        ):
            raise PolicyError("policy identity fields must not be empty")
        graph_node_ids = [node.node_id for node in self.nodes]
        node_ids = [*graph_node_ids, *(node.node_id for node in self.ability_plan)]
        if len(set(node_ids)) != len(node_ids):
            raise PolicyError("policy node IDs must be unique")
        if any(node.kind != NodeKind.ABILITY for node in self.ability_plan):
            raise PolicyError("ability plan may contain only ability nodes")
        claim_ids = [claim.claim_id for claim in self.evidence]
        if len(set(claim_ids)) != len(claim_ids):
            raise PolicyError("policy evidence IDs must be unique")
        if self.entry not in set(graph_node_ids):
            raise PolicyError("policy entry does not resolve")
        for claim in self.evidence:
            if claim.snapshot_id != self.snapshot_id:
                raise PolicyError(f"claim {claim.claim_id} uses a stale snapshot")
        _validate_counter_cards(self.counter_cards, self.nodes, claim_ids)
        _validate_core_alternative_cards(self.core_alternatives, self.nodes, claim_ids)
        if self.schema_version == 1 and self.core_alternatives:
            raise PolicyError("policy schema 1 cannot contain core alternatives")

    @property
    def policy_id(self) -> str:
        return sha256_json(self.as_dict(include_policy_id=False))

    def as_dict(self, *, include_policy_id: bool = True) -> dict[str, object]:
        from .policy_codec import unstructure_build_policy

        return unstructure_build_policy(self, include_policy_id=include_policy_id)

    @classmethod
    def from_dict(cls, value: dict[str, object]) -> BuildPolicy:
        """Decode a policy sidecar and verify its fingerprint.

        Returns:
            A fully typed policy graph.

        Raises:
            PolicyError: If structure, enums, identity, or fingerprint are invalid.

        """
        from .policy_codec import structure_build_policy

        return structure_build_policy(value)

File: crates/deadlock-guides/src/policy_model.rs
use std::collections::BTreeSet;

use deadlock_data::{Error, Result, fingerprint, object};
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::policy_cards::{CoreAlternativeCard, CounterCard};
use crate::policy_claim::EvidenceClaim;
use crate::policy_guard::BranchCondition;
use crate::policy_node::{NodeKind, PolicyNode};

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub enum AbstentionReason {
    #[serde(rename = "stale_or_incomplete_mechanics")]
    StaleMechanics,
    #[serde(rename = "inadequate_support_or_overlap")]
    InadequateSupport,
    #[serde(rename = "evidence_conflict")]
    EvidenceConflict,
    #[serde(rename = "illegal_path")]
    IllegalPath,
    #[serde(rename = "unclear_threat")]
    UnclearThreat,
    #[serde(rename = "out_of_distribution_state")]
    OutOfDistribution,
    #[serde(rename = "telemetry_failure")]
    TelemetryFailure,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Abstention {
    pub reason: AbstentionReason,
    pub detail: String,
    #[serde(default)]
    pub node_id: Option<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct BuildPolicyContent {
    pub schema_version: u8,
    pub hero_id: u64,
    pub variant: String,
    pub invariant_kit_id: String,
    pub strategic_role: String,
    pub snapshot_id: String,
    pub entry: String,
    pub nodes: Vec<PolicyNode>,
    pub evidence: Vec<EvidenceClaim>,
    #[serde(default)]
    pub ability_plan: Vec<PolicyNode>,
    #[serde(default)]
    pub abstentions: Vec<Abstention>,
    #[serde(default)]
    pub counter_cards: Vec<CounterCard>,
    #[serde(default)]
    pub core_alternatives: Vec<CoreAlternativeCard>,
    #[serde(default = "default_path_id")]
    pub path_id: String,
    #[serde(default = "default_path_label")]
    pub path_label: String,
}

fn default_path_id() -> String {
    "default".into()
}
fn default_path_label() -> String {
    "Evidence Default".into()
}

#[derive(Clone, Debug)]
pub struct BuildPolicy {
    content: BuildPolicyContent,
    policy_id: String,
}

impl BuildPolicy {
    /// # Errors
    /// Returns an error when policy identity, nodes, claims, or evidence cards fail structural validation.
    pub fn new(mut content: BuildPolicyContent) -> Result<Self> {
        normalize_branches(&mut content);
        validate_identity(&content)?;
        validate_claims(&content)?;
        validate_cards(&content)?;
        let policy_id = fingerprint(&serde_json::to_value(&content)?)?;
        Ok(Self { content, policy_id })
    }

    /// # Errors
    /// Returns an error when the sidecar has unknown fields, invalid values, or an incorrect fingerprint.
    pub fn from_document(value: &Value) -> Result<Self> {
        let mut payload = object(value)?.clone();
        let expected = payload.remove("policy_id").filter(|value| !value.is_null());
        let expected = expected
            .as_ref()
            .map(|value| {
                value
                    .as_str()
                    .ok_or_else(|| Error::new("Policy fingerprint must be a string"))
            })
            .transpose()?;
        let policy = Self::new(serde_json::from_value(payload.into())?)?;
        if expected.is_some_and(|expected| expected != policy.policy_id) {
            return Err(Error::new("Policy fingerprint does not match its contents"));
        }
        Ok(policy)
    }

    #[must_use]
    pub const fn content(&self) -> &BuildPolicyContent {
        &self.content
    }

    #[must_use]
    pub fn policy_id(&self) -> &str {
        &self.policy_id
    }

    /// # Errors
    /// Returns an error when JSON serialization fails.
    pub fn to_document(&self, include_policy_id: bool) -> Result<Value> {
        let mut value = serde_json::to_value(&self.content)?;
        if include_policy_id {
            value["policy_id"] = self.policy_id.clone().into();
        }
        Ok(value)
    }
}

fn normalize_branches(content: &mut BuildPolicyContent) {
    for node in content.nodes.iter_mut().chain(&mut content.ability_plan) {
        for branch in &mut node.branches {
            if let BranchCondition::All(guards) = &mut branch.when
                && guards.len() == 1
                && let Some(guard) = guards.pop()
            {
                branch.when = BranchCondition::Guard(guard);
            }
        }
    }
}

fn validate_identity(content: &BuildPolicyContent) -> Result<()> {
    if !(1..=5).contains(&content.schema_version) || content.hero_id == 0 {
        return Err(Error::new(
            "Policy has an unsupported schema or hero identifier",
        ));
    }
    let identity = [
        &content.variant,
        &content.invariant_kit_id,
        &content.strategic_role,
        &content.snapshot_id,
        &content.entry,
        &content.path_id,
        &content.path_label,
    ];
    if identity.iter().any(|value| value.trim().is_empty()) {
        return Err(Error::new("Policy identity fields must not be empty"));
    }
    let mut node_ids = BTreeSet::new();
    for node in content.nodes.iter().chain(&content.ability_plan) {
        node.validate()?;
        if !node_ids.insert(&node.node_id) {
            return Err(Error::new("Policy node identifiers must be unique"));
        }
    }
    if !content
        .nodes
        .iter()
        .any(|node| node.node_id == content.entry)
    {
        return Err(Error::new("Policy entry does not resolve"));
    }
    if content
        .ability_plan
        .iter()
        .any(|node| node.kind != NodeKind::Ability)
    {
        return Err(Error::new("Ability plan can contain only ability nodes"));
    }
    if content
        .abstentions
        .iter()
        .any(|row| row.detail.trim().is_empty())
    {
        return Err(Error::new("Abstention detail must not be empty"));
    }
    Ok(())
}

fn validate_claims(content: &BuildPolicyContent) -> Result<()> {
    let mut ids = BTreeSet::new();
    for claim in &content.evidence {
        claim.validate()?;
        if !ids.insert(&claim.claim_id) || claim.snapshot_id != content.snapshot_id {
            return Err(Error::new(
                "Policy claim has a duplicate identifier or stale snapshot",
            ));
        }
    }
    Ok(())
}

fn validate_cards(content: &BuildPolicyContent) -> Result<()> {
    let claims = content
        .evidence
        .iter()
        .map(|claim| claim.claim_id.as_str())
        .collect::<BTreeSet<_>>();
    let actions = content
        .nodes
        .iter()
        .filter(|node| node.kind == NodeKind::Purchase && node.optional)
        .map(|node| (node.item_id, node.evidence_ref.as_deref()))
        .collect::<BTreeSet<_>>();
    let mut ids = BTreeSet::new();
    for card in &content.counter_cards {
        card.validate()?;
        if !ids.insert(card.item_id)
            || !claims.contains(card.evidence_ref.as_str())
            || !actions.contains(&(Some(card.item_id), Some(card.evidence_ref.as_str())))
        {
            return Err(Error::new(
                "Counter card has duplicate items or missing evidence and optional actions",
            ));
        }
    }
    validate_alternatives(content, &claims)
}

fn validate_alternatives(content: &BuildPolicyContent, claims: &BTreeSet<&str>) -> Result<()> {
    if content.schema_version == 1 && !content.core_alternatives.is_empty() {
        return Err(Error::new(
            "Policy schema 1 cannot contain core alternatives",
        ));
    }
    let default = content
        .nodes
        .iter()
        .filter(|node| node.kind == NodeKind::Purchase && !node.optional)
        .filter_map(|node| node.item_id)
        .collect::<BTreeSet<_>>();
    let mut ids = BTreeSet::new();
    for card in &content.core_alternatives {
        card.validate()?;
        if !ids.insert(card.item_id)
            || !claims.contains(card.evidence_ref.as_str())
            || default.contains(&card.item_id)
            || !default.contains(&card.comparator_item_id)
        {
            return Err(Error::new(
                "Core alternative card has duplicate items, missing evidence, or an invalid replacement",
            ));
        }
    }
    Ok(())
}
